### PR TITLE
Add pagination

### PR DESCRIPTION
**Pull request by ellabuysse**
Added pagination with spinner while cards are loading (can't add UI frames in background thread) . Also fixed collection view image bug by adding async call to get the image data before setting the image.

### CODE REVIEW

File: RecipeMatch/Views/DraggableView.h [hunk 1]
```diff
@@ -56,6 +56,6 @@
 @property (nonatomic,strong)UILabel* saveLabel;
 @property (nonatomic,strong)UILabel* saveCount;
 
-- (void)leftClickAction;
-- (void)rightClickAction;
+- (void)leftClickAction:(void (^)(BOOL succeeded, NSError *error))completion;
+- (void)rightClickAction:(void (^)(BOOL succeeded, NSError *error))completion;
```
Comment (michaeltabfb) [inline RecipeMatch/Views/DraggableView.h:60]: nice! i love seeing how you understand the power of completion blocks and when you can use it!💪🏻

File: RecipeMatch/Views/DraggableViewBackground.m [hunk 2]
```diff
@@ -136,7 +142,12 @@ - (void)swipeRight{
     [UIView animateWithDuration:0.2 animations:^{
         dragView.overlayView.alpha = 1;
     }];
-    [dragView rightClickAction];
+    [dragView rightClickAction:^(BOOL succeeded, NSError *error){  // check index after card is dismissed from view
+        if(succeeded){
+            self->allCardsIndex++;
+            [self checkCardIndexStatus];
+        }
+    }];
```
Comment (michaeltabfb) [inline RecipeMatch/Views/DraggableViewBackground.m:150]: i noticed that you call this `rightClickAction ` from `DraggableViewBackground`, then inside `rightClickAction` (that is in `DraggableView`) you call `delegate draggableViewCardSwipedRight`, which is in.... `DraggableViewBackground`..... do you see where i'm going with it?

File: RecipeMatch/Views/DraggableViewBackground.m [hunk 3]
```diff
@@ -18,6 +18,7 @@ @interface DraggableViewBackground ()
 @implementation DraggableViewBackground{
     NSInteger cardsLoadedIndex; // the index of the last card loaded into the loadedCards array
     NSMutableArray *loadedCards; // the array of card loaded
+    NSInteger allCardsIndex; // index of current card out of all cards
```
Comment (michaeltabfb) [inline RecipeMatch/Views/DraggableViewBackground.m:21]: i find the naming of `cardsLoadedIndex ` and `allCardsIndex ` to be little confusing....any chance you can comeout with a better naming?

File: RecipeMatch/Views/GridRecipeCell.h [hunk 4]
```diff
@@ -12,6 +12,7 @@ NS_ASSUME_NONNULL_BEGIN
 @interface GridRecipeCell : UICollectionViewCell
 @property (weak, nonatomic) IBOutlet UILabel *recipeTitle;
 @property (weak, nonatomic) IBOutlet UIImageView *imageView;
+@property (strong, nonatomic) NSString *hasImage;
```
Comment (michaeltabfb) [inline RecipeMatch/Views/GridRecipeCell.h:15]: doesn't seems to be in use

**Review by michaeltabfb — CHANGES_REQUESTED** [review]
So in general, this logic works!👍🏻
as i mentioned in the comments, kudos for start thinking how completion blocks can help you code better!💪🏻

few things though
- seems like this logic isn't implemented for when actually swiping the recipes to either of the sides (working when tapping the X\V buttons)
- this is pretty basic implementation of pagination. ideally you will not want the user the see a loader when loading a new page, but would want to have a smoother user experience. this is up to you what you want to do with this. again, it's working now, but i'd say it's not ideal - let me know if you want to brainstorm about ideas on how to achieve a smoother user experience here.

other then that, there are just few fixes that i'd like you to fix before we can merge it.

File: RecipeMatch/View Controllers/ProfileViewController.m [hunk 5]
```diff
@@ -89,9 +89,25 @@ - (NSInteger)collectionView:(nonnull UICollectionView *)collectionView numberOfI
...
+                dispatch_async(dispatch_get_main_queue(), ^{
+                    // get correct cell for index path
+                    GridRecipeCell *updateCell = (id)[collectionView cellForItemAtIndexPath:indexPath];
+                    if (updateCell)
+                        updateCell.imageView.image = image;
+                });
+            }
+        }
+    }];
+    [task resume];
```
[the remark this answers is not in the document]
Comment (ellabuysse) [reply RecipeMatch/View Controllers/ProfileViewController.m]: I found some images weren't loading so I added this to get the image data before setting the image

File: RecipeMatch/View Controllers/ProfileViewController.m [hunk 6]
```diff
@@ -89,9 +89,25 @@ - (NSInteger)collectionView:(nonnull UICollectionView *)collectionView numberOfI
...
-    [cell.imageView setImageWithURL:[NSURL URLWithString:imageUrl]];
+    NSURL *url = [NSURL URLWithString:imageUrl];
+    NSURLSessionTask *task = [[NSURLSession sharedSession] dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+        if (data) {
+            // load image data asynchronously
+            UIImage *image = [UIImage imageWithData:data];
+            if (image) {
+                dispatch_async(dispatch_get_main_queue(), ^{
+                    // get correct cell for index path
+                    GridRecipeCell *updateCell = (id)[collectionView cellForItemAtIndexPath:indexPath];
```
[the remark this answers is not in the document]
Comment (ellabuysse) [reply RecipeMatch/View Controllers/ProfileViewController.m]: Added a comment for this

File: RecipeMatch/View Controllers/ProfileViewController.m [hunk 7]
```diff
@@ -87,38 +88,18 @@ - (NSInteger)collectionView:(nonnull UICollectionView *)collectionView numberOfI
...
-        }
-    }];
-    [task resume];
-    completion(NO, nil);
+// called by cellForItemAtIndexPath to set up cell with image and title
+- (GridRecipeCell *)setupCell:(GridRecipeCell *)cell withRecipe:(SavedRecipe *)recipe {
+    [cell.imageView sd_setImageWithURL:[NSURL URLWithString:recipe[@"image"]] placeholderImage:[UIImage systemImageNamed:@"photo"]];
+    cell.imageView.layer.cornerRadius = CORNER_RADIUS;
+    cell.recipeTitle.text = recipe[@"name"];
+    return cell;
```
Comment (michaeltabfb) [inline RecipeMatch/View Controllers/ProfileViewController.m:96]: few things.
- `recipe` is already a model object, so you can access it's image\name with `recipe.name`
-  `cell` is being passed to here (and not a copy), so no need to return it back
-  it's better to move this setup functionality to the cell it self, and do all of this from the there, and in `cellForItemAtIndexPath ` just call `[cell setupWithRecipe:self.recipes[indexPath.row]]`

File: RecipeMatch/API/APIManager.m [hunk 8]
```diff
@@ -96,7 +96,7 @@ - (void)getRecipesWithPreferences:(NSString * _Nullable)preferences andCompletio
     NSURL *url = [NSURL URLWithString:apiString];
     [self requestFromAPIWithURL:url andCompletion:^(NSDictionary *dataDictionary, NSError *error) {
         int count = [dataDictionary[@"count"] integerValue];
-        if (count > MIN_RECIPE_COUNT) {
+        if (count > MIN_RECIPE_COUNT) {  // if there are enough recipes, return them
```
Comment (michaeltabfb) [inline RecipeMatch/API/APIManager.m:99]: it might be worth adding more description here, like the logic and actual reasoning for it

**Review by michaeltabfb — APPROVED** [review]
looks good.
i left few comments which i think would be worth visiting and iterating on

File: RecipeMatch/View Controllers/StreamViewController.m [hunk 9]
```diff
@@ -35,54 +35,55 @@ - (void)viewDidLoad {
...
     spinner.center = CGPointMake(self.view.center.x, self.view.center.y);
     [self.view addSubview:spinner];
     [spinner startAnimating];
     [self getRecipesWithPreferencesWithCompletion:^(NSArray *recipes, NSError * _Nullable error){
-        if(recipes){
+        if (recipes) {
             self.draggableBackground = [[DraggableViewBackground alloc]initWithFrame:self.view.frame];
             self.draggableBackground.delegate = self;
-            self.draggableBackground.recipes = (NSMutableArray *)recipes;
+            self.draggableBackground.recipes = recipes;
```
[the remark this answers is not in the document]
Comment (michaeltabfb) [reply RecipeMatch/View Controllers/StreamViewController.m:52]: this actually needs to be fixed, otherwise might lead to a crash

**Comment by michaeltabfb** [comment]
also, now that you added the sdWebImage library, you might want to change all other image loading methods to also use it. that way you'll get it benefits like caching and a lot of others

File: RecipeMatch/Views/GridRecipeCell.m [hunk 10]
```diff
@@ -9,11 +9,19 @@
 #import "UIKit+AFNetworking.h"
 
 @implementation GridRecipeCell
+static const float CORNER_RADIUS = 15;
+
 // clears image and title of cell
 - (void)prepareForReuse {
     [super prepareForReuse];
     [self.imageView cancelImageDownloadTask];
```
Comment (michaeltabfb) [inline RecipeMatch/Views/GridRecipeCell.m:17]: might as well just use the sdWebImage version for canceling `sd_cancelCurrentImageLoad`

**Pull request merged**